### PR TITLE
Task - Implement multi-level dialog hierarchy for the lesson plan

### PR DESCRIPTION
**Pull request by jsy1218-zz**
lesson plan construction at runtine without having to modify the source code. This is the intial check in for the prototype.

### CODE REVIEW

File: DTML.EduBot/LessonPlan/Lesson.cs
Comment (eklavyamirani) [inline DTML.EduBot/LessonPlan/Lesson.cs]: Do we also want to store progress here?

File: DTML.EduBot/LessonPlan/LessonPlan.cs
Comment (eklavyamirani) [inline DTML.EduBot/LessonPlan/LessonPlan.cs]: Progress here too?

File: DTML.EduBot/Dialogs/RootDialog.cs
Comment (eklavyamirani) [inline DTML.EduBot/Dialogs/RootDialog.cs]: Let's move these conversational strings out into constants.

File: DTML.EduBot/Dialogs/RootDialog.cs
Comment (eklavyamirani) [inline DTML.EduBot/Dialogs/RootDialog.cs]: Success, fail -> enums

File: DTML.EduBot/Dialogs/RootDialog.cs
Comment (eklavyamirani) [inline DTML.EduBot/Dialogs/RootDialog.cs]: This looks great!

File: DTML.EduBot/Dialogs/RootDialog.cs
Comment (eklavyamirani) [inline DTML.EduBot/Dialogs/RootDialog.cs]: Why did we need a container to resolve the dependency for LessonPlanDialog here? Can you inject this dependency from constructor instead?

File: DTML.EduBot/Dialogs/LessonPlanDialog.cs
Comment (eklavyamirani) [inline DTML.EduBot/Dialogs/LessonPlanDialog.cs]: Possible null reference. Lessons could be null. (Lesson plans Hard coded now, but it will be changed later)

File: DTML.EduBot/Dialogs/LessonDialog.cs
Comment (eklavyamirani) [inline DTML.EduBot/Dialogs/LessonDialog.cs]: possible null reference exceptions?

File: DTML.EduBot/Dialogs/LessonDialog.cs
Comment (eklavyamirani) [inline DTML.EduBot/Dialogs/LessonDialog.cs]: Can you move out this hard coded stuff and store it as data in this.lesson, so the text would look like -> 

Text = lesson.topic.questions[i].text (example)

Let's work with Soumya to build a lesson plan object for her lesson plans.

File: DTML.EduBot/Dialogs/LessonDialog.cs
Comment (eklavyamirani) [inline DTML.EduBot/Dialogs/LessonDialog.cs]: can you either move the using statements inside or outside the namespace? currently they are mixed.

File: DTML.EduBot/Dialogs/GreetingDialog.cs
Comment (eklavyamirani) [inline DTML.EduBot/Dialogs/GreetingDialog.cs]: ChitChat dialog?

File: DTML.EduBot/Controllers/MessagesController.cs
Comment (eklavyamirani) [inline DTML.EduBot/Controllers/MessagesController.cs]: use the BotName constant instead of "Edubot"

File: DTML.EduBot/Controllers/MessagesController.cs
Comment (eklavyamirani) [inline DTML.EduBot/Controllers/MessagesController.cs]: Async methods would generally have an async suffix. Can you rename HandleSystemMessage to HandleSystemMessageAsync?

**Review by eklavyamirani — CHANGES_REQUESTED** [review]
Great job!

Looks good overall. Thank you for your hard work!

File: DTML.EduBot/Dialogs/RootDialog.cs
Comment (jsy1218-zz) [reply DTML.EduBot/Dialogs/RootDialog.cs]: I wasn't aware of DI from ctor. Let me research on how to do it.

File: DTML.EduBot/LessonPlan/Lesson.cs
Comment (jsy1218-zz) [reply DTML.EduBot/LessonPlan/Lesson.cs]: I'm still thinking how to implement storing progress..

File: DTML.EduBot/Dialogs/RootDialog.cs
Comment (jsy1218-zz) [reply DTML.EduBot/Dialogs/RootDialog.cs]: I decided not to declare lessonPlanDialog as a field, because it was used once in the class.

File: DTML.EduBot/Dialogs/GreetingDialog.cs
Comment (jsy1218-zz) [reply DTML.EduBot/Dialogs/GreetingDialog.cs]: Modified.

File: DTML.EduBot/Dialogs/LessonDialog.cs
Comment (jsy1218-zz) [reply DTML.EduBot/Dialogs/LessonDialog.cs]: inside the namespace now.

File: DTML.EduBot/Dialogs/LessonDialog.cs
Comment (jsy1218-zz) [reply DTML.EduBot/Dialogs/LessonDialog.cs]: Yup, now it reads from the lesson plan file.

File: DTML.EduBot/Controllers/MessagesController.cs
Comment (jsy1218-zz) [reply DTML.EduBot/Controllers/MessagesController.cs]: Modified.

File: DTML.EduBot/Controllers/MessagesController.cs
Comment (jsy1218-zz) [reply DTML.EduBot/Controllers/MessagesController.cs]: Yup

File: DTML.EduBot/Dialogs/RootDialog.cs
Comment (jsy1218-zz) [reply DTML.EduBot/Dialogs/RootDialog.cs]: I actually changed the type from string to bool

File: DTML.EduBot/Dialogs/LessonDialog.cs
Comment (jsy1218-zz) [reply DTML.EduBot/Dialogs/LessonDialog.cs]: It shouldn't, because user has to type something to trigger the method delegate CheckAnswerAsync

File: DTML.EduBot/Dialogs/LessonPlanDialog.cs
Comment (jsy1218-zz) [reply DTML.EduBot/Dialogs/LessonPlanDialog.cs]: I'm thinking instead of null check, we should enforce the lesson to be present in the lesson plan.

File: DTML.EduBot/Dialogs/RootDialog.cs
Comment (jsy1218-zz) [reply DTML.EduBot/Dialogs/RootDialog.cs]: Do we want to do it now, or later, for addressing the style?

File: DTML.EduBot/Dialogs/RootDialog.cs
Comment (eklavyamirani) [reply DTML.EduBot/Dialogs/RootDialog.cs]: Well, it is still a dependency. This is kind of an [anti-pattern](http://blog.ploeh.dk/2010/02/03/ServiceLocatorisanAnti-Pattern/).

If it is not too much work, I'd suggest we move it out. Can you do it as a follow up later? I will not block the PR for this.

File: DTML.EduBot/Dialogs/LessonPlanDialog.cs
Comment (eklavyamirani) [reply DTML.EduBot/Dialogs/LessonPlanDialog.cs]: either way, I would put it at both places honestly.

File: DTML.EduBot/Dialogs/LessonDialog.cs
Comment (eklavyamirani) [reply DTML.EduBot/Dialogs/LessonDialog.cs]: okay.

File: DTML.EduBot/Controllers/MessagesController.cs
Comment (eklavyamirani) [inline DTML.EduBot/Controllers/MessagesController.cs:26]: Followup: please use constructor injection.

File: DTML.EduBot/Dialogs/BasicDialogModule.cs
Comment (eklavyamirani) [inline DTML.EduBot/Dialogs/BasicDialogModule.cs:11]: This is where the code would go for Constructor injection. You would need to replace the dependency injection in web api to use this resolver.

File: DTML.EduBot/Dialogs/ChitChatDialog.cs
Comment (eklavyamirani) [inline DTML.EduBot/Dialogs/ChitChatDialog.cs]: Remove keys before checking in please.

File: DTML.EduBot/Dialogs/RootDialog.cs
Comment (eklavyamirani) [inline DTML.EduBot/Dialogs/RootDialog.cs:54]: I might be missing something, but did you mention that you were using a bool now?

File: DTML.EduBot/Dialogs/RootDialog.cs
Comment (eklavyamirani) [inline DTML.EduBot/Dialogs/RootDialog.cs:80]: If else? :) You can change it as a follow up if you want.

Edit: I am approving the code, but please change this before merging.

**Review by eklavyamirani — CHANGES_REQUESTED** [review]
Remove LUIS and QNA keys before checking in.

Also follow up: use constructor injection instead of resolvers in code. We can talk about it if you want, and I can explain why. This [blog](http://blog.ploeh.dk/2010/02/03/ServiceLocatorisanAnti-Pattern/) explains it well.

**Review by eklavyamirani — APPROVED** [review]

**Comment by eklavyamirani** [comment]
@jsy1218 Please look at #7 and #8

**Pull request merged**
